## Importing libs


In [11]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import os
import requests
import tweepy 
import json
#from PIL import Image
from io import BytesIO

## Gathering Data
- csv file (twitter_archive_enhanced.csv)

In [12]:
df_archive = pd.read_csv('twitter_archive_enhanced.csv')


In [13]:
df_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

- Image predictions

In [14]:
# Download the image prediction file programmatically
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
image_request = requests.get(url, allow_redirects=True)
open('image_predictions.tsv', 'wb').write(image_request.content)
df_image = pd.read_csv('image_predictions.tsv', sep = '\t')
df_image.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

- Twitter API data

In [10]:
#I used twitter_api.py from Udacity Supporting Materials
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = df_1.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

NameError: name 'df_1' is not defined

- tweet_json.txt

In [15]:
#get the data from the tweet_json.txt file and convert it to a list
df_text_list=[]
with open('tweet_json.txt',encoding='utf-8') as file:
    for item in file:
        df_text_list.append(json.loads(item))
        
df_text_list[100]

{'created_at': 'Thu Jun 08 04:17:07 +0000 2017',
 'id': 872668790621863937,
 'id_str': '872668790621863937',
 'full_text': 'RT @loganamnosis: Penelope here is doing me quite a divertir. Well done, @dog_rates! Loving the pupdate. 14/10, je jouerais de nouveau. htt…',
 'truncated': False,
 'display_text_range': [0, 140],
 'entities': {'hashtags': [],
  'symbols': [],
  'user_mentions': [{'screen_name': 'loganamnosis',
    'name': 'michael',
    'id': 154767397,
    'id_str': '154767397',
    'indices': [3, 16]},
   {'screen_name': 'dog_rates',
    'name': 'WeRateDogs™ (author)',
    'id': 4196983835,
    'id_str': '4196983835',
    'indices': [73, 83]}],
  'urls': []},
 'source': '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
 'in_reply_to_status_id': None,
 'in_reply_to_status_id_str': None,
 'in_reply_to_user_id': None,
 'in_reply_to_user_id_str': None,
 'in_reply_to_screen_name': None,
 'user': {'id': 4196983835,
  'id_str': '4196983835',
  'name

In [16]:
# saving the text list as a dataframe
df_text=pd.DataFrame(df_text_list,columns=['id','retweet_count','favorite_count'])
df_text.head()

id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

In [17]:
#changing id column name into tweet_id to be like another dataframes
df_text.rename(columns={'id':'tweet_id'},inplace=True)
df_text.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

## Assessing Data

- Twitter archive DataFrame

In [18]:
df_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [19]:
df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [20]:
df_archive.isna().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [21]:
df_archive.tweet_id.duplicated().sum()

0

In [22]:
df_archive.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [23]:
df_archive[df_archive.rating_numerator<10].tweet_id

45      883482846933004288
229     848212111729840128
315     835152434251116546
387     826598799820865537
462     817502432452313088
485     814578408554463233
599     798682547630837760
605     798576900688019456
730     781661882474196992
745     780092040432480260
764     777953400541634568
765     777885040357281792
784     775096608509886464
814     771014301343748096
835     768193404517830656
845     766423258543644672
859     763183847194451968
860     763167063695355904
883     760252756032651264
896     759099523532779520
906     758041019896193024
912     757596066325864448
915     757354760399941633
936     753375668877008896
946     752568224206688256
948     752334515931054080
956     751583847268179968
964     750506206503038976
966     750383411068534784
987     749036806121881602
               ...        
2312    666776908487630848
2314    666701168228331520
2315    666691418707132416
2316    666649482315059201
2317    666644823164719104
2319    666447344410484738
2

In [24]:
df_archive[df_archive.rating_numerator<10].count()[0]

440

In [25]:
df_archive.name.value_counts()

None        745
a            55
Charlie      12
Cooper       11
Oliver       11
Lucy         11
Lola         10
Tucker       10
Penny        10
Winston       9
Bo            9
Sadie         8
the           8
Bailey        7
Toby          7
Buddy         7
an            7
Daisy         7
Milo          6
Scout         6
Bella         6
Stanley       6
Rusty         6
Koda          6
Oscar         6
Jack          6
Dave          6
Jax           6
Leo           6
Alfie         5
           ... 
Cupid         1
Ralf          1
Stefan        1
Keet          1
Ricky         1
Lance         1
Sparky        1
Tessa         1
Luther        1
Boston        1
Bobby         1
Tobi          1
Carper        1
Kenzie        1
Autumn        1
Stewie        1
Callie        1
Gòrdón        1
Jennifur      1
Ronnie        1
Gin           1
Gilbert       1
Kanu          1
Caryl         1
Buckley       1
Gunner        1
Skittle       1
Ralpher       1
Jessiga       1
Leonidas      1
Name: name, Length: 957,

- Image predictions Dataframe

In [26]:
df_image

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [27]:
# cheack the Urls are working !
# the image of tweet_id:666020888022790149
from IPython.display import Image
Image(url='https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg')

In [28]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [29]:
df_image.isna().sum()

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

In [30]:
df_image.jpg_url.duplicated().sum()

66

In [31]:
df_image.jpg_url.isna().sum()

0

In [32]:
df_image.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [33]:
pd.concat(i for _, i in df_image.groupby("jpg_url") if len(i) > 1)

tweet_id                                            jpg_url  \
480   675354435921575936  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1641  807106840509214720  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1864  842892208864923648  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1691  815966073409433600  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1858  841833993020538882  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1715  819004803107983360    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1716  819006400881917954    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1738  822244816520155136    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1740  822489057087389700    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1767  826958653328592898    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1789  829878982036299777    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1785  829374341691346946    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1903  851953902622658560    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1791  830583320585068544    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
224   670319130621435904    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
241   670444955656130560    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
1590  798644042770751489    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
85    667509364010450944    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
1589  798628517273620480    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
327   671896809300709376    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
1591  798665375516884993    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
...                  ...                                                ...   
1530  789530877013393408    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg   
1571  794983741416415232    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg   
1533  789986466051088384    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1538  790723298204217344    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1534  790277117346975746    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg   
1683  813944609378369540    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg   
1539  790946055508652032    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1544  791672322847637504    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1564  793614319594401792    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1561  793286476301799424    https://pbs.twimg.com/media/CwJR1okWIAA6XMp.jpg   
1804  832215726631055365    https://pbs.twimg.com/media/CwJR1okWIAA6XMp.jpg   
1566  793962221541933056    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1652  809808892968534016    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1572  795076730285391872    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg   
1755  824796380199809024    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg   
1578  796149749086875649    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg   
1579  796177847564038144    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg   
1610  801167903437357056    https://pbs.twimg.com/media/Cx5R8wPVEA

In [34]:
print(df_image.p1_dog.value_counts())
print(df_image.p2_dog.value_counts())
print(df_image.p3_dog.value_counts())

True     1532
False     543
Name: p1_dog, dtype: int64
True     1553
False     522
Name: p2_dog, dtype: int64
True     1499
False     576
Name: p3_dog, dtype: int64


- Text file

In [35]:
df_text

tweet_id  retweet_count  favorite_count
0     892420643555336193           8853           39467
1     892177421306343426           6514           33819
2     891815181378084864           4328           25461
3     891689557279858688           8964           42908
4     891327558926688256           9774           41048
5     891087950875897856           3261           20562
6     890971913173991426           2158           12041
7     890729181411237888          16716           56848
8     890609185150312448           4429           28226
9     890240255349198849           7711           32467
10    890006608113172480           7624           31166
11    889880896479866881           5156           28268
12    889665388333682689           8538           38818
13    889638837579907072           4735           27672
14    889531135344209921           2321           15359
15    889278841981685760           5637           25652
16    888917238123831296           4709           29611
17    888804989199671297           4559           26080
18    888554962724278272           3732           20290
19    888078434458587136           3653           22201
20    887705289381826560           5609           30779
21    887517139158093824          12082           46959
22    887473957103951883          18781           69871
23    887343217045368832          10737           34222
24    887101392804085760           6167           31061
25    886983233522544640           8084           35859
26    886736880519319552           3443           12306
27    886680336477933568           4610           22798
28    886366144734445568           3316           21524
29    886267009285017600              4             117
...                  ...            ...             ...
2324  666411507551481857            339             459
2325  666407126856765440             44             113
2326  666396247373291520             92             172
2327  666373753744588802            100             194
2328  666362758909284353            595             804
2329  666353288456101888             77             229
2330  666345417576210432            146             307
2331  666337882303524864             96             204
2332  666293911632134144            368             522
2333  666287406224695296             71             152
2334  666273097616637952             82             184
2335  666268910803644416             37             108
2336  666104133288665088           6871           14765
2337  666102155909144576             16              81
2338  666099513787052032             73             164
2339  666094000022159362             79             169
2340  666082916733198337             47             121
2341  666073100786774016            174             335
2342  666071193221509120             67             154
2343  666063827256086533            232             496
2344  666058600524156928             61             115
2345  666057090499244032            146             304
2346  666055525042405380            261             448
2347  666051853826850816            879            1253
2348  666050758794694657             60             136
2349  666049248165822465             41             111
2350  666044226329800704            147             311
2351  666033412701032449             47             128
2352  666029285002620928             48             132
2353  666020888022790149            532            2535

[2354 rows x 3 columns]

In [36]:
df_text.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null int64
retweet_count     2354 non-null int64
favorite_count    2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


In [37]:
df_text.isna().sum()

tweet_id          0
retweet_count     0
favorite_count    0
dtype: int64

In [38]:
df_text.duplicated().sum()

0

## DATA IUSSES
### Tidiness
- Data are divied into three dataframes (csv,tsv,test)
- dog stage divided ino 4 columns :doggo, puppo, pupper, floofer 
### Quality 
    1- Archive dataframe:
        - there are many useless columns and NAN columns.
        - combining the columns doggo, puppo, pupper, floofer into a single column
        - worng data types ex:timestarp.
        - correct numerators and denomerator float rate.
        - some denomerator rating column not =10.

    2- Image dataframe:
        - massing rows in image dataframe has only 2075 and archive as 2356.
        - many coulms are useless (well be droped).
        - 66 jpg_url duplicated.
        - some names are lowercae and some are upper case.
        - tweet_id datatype is wrong.
    3- text dataframe:
        - massing rows in text dataframe has only 2354 and archive as 2356

## Clean Archive

In [39]:
#Copping dataframes to clean it
clean_df_archive= df_archive.copy()

### Define
    - drop usless columns
### code  

In [40]:
clean_df_archive.drop(['source','in_reply_to_status_id','in_reply_to_user_id','retweeted_status_user_id'
                    ,'retweeted_status_timestamp','expanded_urls'], axis=1,inplace=True) 
                                                   

### Test

In [41]:
clean_df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id               2356 non-null int64
timestamp              2356 non-null object
text                   2356 non-null object
retweeted_status_id    181 non-null float64
rating_numerator       2356 non-null int64
rating_denominator     2356 non-null int64
name                   2356 non-null object
doggo                  2356 non-null object
floofer                2356 non-null object
pupper                 2356 non-null object
puppo                  2356 non-null object
dtypes: float64(1), int64(3), object(7)
memory usage: 202.5+ KB


### Define
    mareg doggo,floofer,pupper and puppo in dog stage and remove them
### Code

In [42]:
#clean_df_archive['dog_stage']=clean_df_archive['text'].str.extract('(doggo|floofer|pupper|puppo)')


In [43]:
# handle none
clean_df_archive.doggo.replace('None', '', inplace=True)
clean_df_archive.floofer.replace('None', '', inplace=True)
clean_df_archive.pupper.replace('None', '', inplace=True)
clean_df_archive.puppo.replace('None', '', inplace=True)

# merge into column
clean_df_archive['dog_stage'] = clean_df_archive.doggo + clean_df_archive.floofer + clean_df_archive.pupper + clean_df_archive.puppo

# handle multiple stages
clean_df_archive.loc[clean_df_archive.dog_stage == 'doggopupper', 'dog_stage'] = 'doggo, pupper'
clean_df_archive.loc[clean_df_archive.dog_stage == 'doggopuppo', 'dog_stage'] = 'doggo, puppo'
clean_df_archive.loc[clean_df_archive.dog_stage == 'doggofloofer', 'dog_stage'] = 'doggo, floofer'

# handle missing values
clean_df_archive.loc[clean_df_archive.dog_stage == '', 'dog_stage'] = np.nan
clean_df_archive.drop(['doggo','floofer','pupper','puppo'],axis=1,inplace=True)

### Test

In [44]:
clean_df_archive.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   rating_numerator  rating_denominator      name dog_stage  
0                13                  10   Phineas       NaN  
1                13                  10     Tilly       NaN  
2                12                  10    Archie       NaN  
3                13                  10     Darla       NaN  
4                12                  10  Franklin       NaN

In [45]:
clean_df_archive.dog_stage.value_counts()

pupper            245
doggo              83
puppo              29
doggo, pupper      12
floofer             9
doggo, floofer      1
doggo, puppo        1
Name: dog_stage, dtype: int64

### Define
    correct tweet_id &timestarp data type&rating
### Code

In [46]:
#change tweet id data type to string
clean_df_archive.tweet_id=clean_df_archive.tweet_id.astype(str)
#change time stamp data type to string
clean_df_archive.timestamp=pd.to_datetime(clean_df_archive.timestamp)
#change rating_numerator rating_denominator type to string data type to float
clean_df_archive[['rating_numerator', 'rating_denominator']] = clean_df_archive[['rating_numerator','rating_denominator']].astype(float)

### Test

In [47]:
clean_df_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 8 columns):
tweet_id               2356 non-null object
timestamp              2356 non-null datetime64[ns]
text                   2356 non-null object
retweeted_status_id    181 non-null float64
rating_numerator       2356 non-null float64
rating_denominator     2356 non-null float64
name                   2356 non-null object
dog_stage              380 non-null object
dtypes: datetime64[ns](1), float64(3), object(4)
memory usage: 147.3+ KB


### Define
    some denomerator rating column not =10.
### Code


In [48]:
clean_df_archive['rating'] = 10 * clean_df_archive['rating_numerator'] / clean_df_archive['rating_denominator'].astype(float)


### Test

In [49]:
clean_df_archive.head()

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   
1  892177421306343426 2017-08-01 00:17:27   
2  891815181378084864 2017-07-31 00:18:03   
3  891689557279858688 2017-07-30 15:58:51   
4  891327558926688256 2017-07-29 16:00:24   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   rating_numerator  rating_denominator      name dog_stage  rating  
0              13.0                10.0   Phineas       NaN    13.0  
1              13.0                10.0     Tilly       NaN    13.0  
2              12.0                10.0    Archie       NaN    12.0  
3              13.0                10.0     Darla       NaN    13.0  
4              12.0                10.0  Franklin       NaN    12.0

### Define
    DECIMAL DOG RATINGS for dog ratings, there are some observations with decimal ratings
### Code

In [50]:
clean_df_archive[clean_df_archive.text.str.contains(r"(\d+\.\d*\/\d+)")][['text', 'rating_numerator']]


/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  """Entry point for launching an IPython kernel.


text  rating_numerator
45    This is Bella. She hopes her smile made you sm...               5.0
340   RT @dog_rates: This is Logan, the Chow who liv...              75.0
695   This is Logan, the Chow who lived. He solemnly...              75.0
763   This is Sophie. She's a Jubilant Bush Pupper. ...              27.0
1689  I've been told there's a slight possibility he...               5.0
1712  Here we have uncovered an entire battalion of ...              26.0

## Clean Image 

In [51]:
clean_df_image= df_image.copy()

### Define
    make all names upper case
### Code

In [ ]:
clean_df_image.p1=clean_df_image.p1.str.title()
clean_df_image.p2=clean_df_image.p2.str.title()
clean_df_image.p3=clean_df_image.p3.str.title()

### Test

In [ ]:
clean_df_image.head()

### Define
    drop dopulaced 66 urls
### Code

In [ ]:
clean_df_image.jpg_url.drop_duplicates(inplace=True)

### Test

In [ ]:
clean_df_image.jpg_url.duplicated().sum()

### Define
     tweet_id datatype is wrong.
### Code

In [ ]:

clean_df_image.tweet_id=clean_df_image.tweet_id.astype(str)

### Test

In [ ]:
clean_df_image.info()

## Clean text

### Define
    tweet_id datatype is wrong.
### Code

In [ ]:
clean_df_text=df_text.copy()
clean_df_text.tweet_id=clean_df_text.tweet_id.astype(str)

## Merging
### code

In [ ]:
clean_df=pd.merge(clean_df_archive,clean_df_text,on='tweet_id',how='left')
clean_df=pd.merge(clean_df,clean_df_image,on='tweet_id',how='left')

### Test

In [ ]:
clean_df.info()

### Define
     drop retweets and kepp orginal tweets
### Code

In [ ]:
clean_df=clean_df[clean_df.retweeted_status_id.isnull()]
#Test
clean_df.info()

### Define
     drop tweets that has no urls
### Code

In [ ]:
clean_df=clean_df[clean_df.jpg_url.notnull()]
#test
clean_df.info()

In [ ]:
#drop retweeted_status_id 
clean_df.drop(['retweeted_status_id'],axis=1,inplace=True)
#test
clean_df.info()

## Storing Data

In [ ]:
clean_df.to_csv('twitter_archive_master.csv', index=False, encoding='utf-8')

In [ ]:
#test
clean_df=pd.read_csv('twitter_archive_master.csv')
clean_df.info()

## Visualization & Analysis


### The common dog stage

In [ ]:
clean_df.dog_stage.value_counts()[:10].plot('bar')
plt.title('common dog stage')
plt.xlabel('dog stage')
plt.ylabel('Count');

### Common dog Names

In [ ]:
clean_df.name.value_counts().sort_values(ascending =False)[2:10].plot('barh')
plt.title(" Common Dogs' Names")
plt.xlabel('Count')
plt.ylabel("Dog's Name");

### Relation between Retweets count& Favourite count

In [ ]:
plt.scatter(clean_df.retweet_count,clean_df.favorite_count)
plt.xlabel('Retweets count')
plt.ylabel('Favourite count')

## Top retweeted & favorite dog

In [ ]:
#top reweete
photo =clean_df.loc[clean_df['retweet_count'].sort_values(ascending =False).index[0]]
r = photo['jpg_url']
Image(url=r)

In [ ]:
#top favorit
photo =clean_df.loc[clean_df['rating'].sort_values(ascending =False).index[0]]
r = photo['jpg_url']
Image(url=r)